<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro"
       xmlns:gazebo="http://www.gazebosim.org/schema/gazebo_urdf"
       name="diffbot">


  <!-- Parámetros reutilizables -->
  <xacro:property name="PI" value="3.1415926535897931"/>
  <xacro:property name="wheel_radius" value="0.035"/> <!-- Radio de rueda (m) -->
  <xacro:property name="wheel_separation" value="0.23"/> <!-- Separación entre ruedas (m) -->

  <!-- Incluir macros -->
  <xacro:include filename="$(find tp1_robot)/urdf/inertial_macros.xacro"/>
  <xacro:include filename="$(find tp1_robot)/urdf/motor.xacro"/>

  <!-- Base link (chasis) -->
  <link name="base_link">
    <visual>
      <geometry>
        <mesh filename="package://tp1_robot/meshes/chassis.stl"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://tp1_robot/meshes/chassis.stl"/>
      </geometry>
    </collision>
    <xacro:cylinder_inertia m="0.8" r="0.1" h="0.05"/>
  </link>

  <!-- Ruedas -->
  <!-- Rueda izquierda -->
  <joint name="left_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="left_wheel_link"/>
    <axis xyz="0 1 0"/>
    <origin xyz="${-wheel_separation/2} -0.08 -0.035" rpy="0 0 0"/>
  </joint>

  <link name="left_wheel_link">
    <visual>
      <geometry>
        <mesh filename="package://tp1_robot/meshes/wheel.stl"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://tp1_robot/meshes/wheel.stl"/>
      </geometry>
    </collision>
    <xacro:cylinder_inertia m="0.15" r="0.035" h="0.025"/>
  </link>

  <!-- Rueda derecha -->
  <joint name="right_wheel_joint" type="continuous">
    <parent link="base_link"/>
    <child link="right_wheel_link"/>
    <axis xyz="0 1 0"/>
    <origin xyz="${wheel_separation/2} -0.08 -0.035" rpy="0 0 0"/>
  </joint>

  <link name="right_wheel_link">
    <visual>
      <geometry>
        <mesh filename="package://tp1_robot/meshes/wheel.stl"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://tp1_robot/meshes/wheel.stl"/>
      </geometry>
    </collision>
    <xacro:cylinder_inertia m="0.15" r="0.035" h="0.025"/>
  </link>

  <!-- Motores (usando macro) -->
  <xacro:motor prefix="left" chassis_width="0.1" reflect="false"/>
  <xacro:motor prefix="right" chassis_width="0.1" reflect="true"/>

  <!-- ROS2 Control -->
  <ros2_control name="diffbot_ros2_control" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>
    <joint name="left_wheel_joint">
      <command_interface name="velocity">
        <param name="min">-10.0</param>
        <param name="max">10.0</param>
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="position"/>
    </joint>
    <joint name="right_wheel_joint">
      <command_interface name="velocity">
        <param name="min">-10.0</param>
        <param name="max">10.0</param>
      </command_interface>
      <state_interface name="velocity"/>
      <state_interface name="position"/>
    </joint>
  </ros2_control>

  <!-- Plugin para Gazebo -->
  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>$(find tp1_robot)/config/diffbot_controllers.yaml</parameters>
    </plugin>
  </gazebo>

</robot>